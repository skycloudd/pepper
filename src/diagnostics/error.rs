use super::{Diag, ErrorSpan};
use crate::span::Span;
use chumsky::error::{Rich, RichReason};
use codespan_reporting::diagnostic::Severity;
use core::fmt::Display;
use owo_colors::OwoColorize;
use std::borrow::Cow;

#[derive(Clone, Debug)]
pub enum Error {
    ExpectedFound {
        expected: Vec<String>,
        found: Option<String>,
        span: Span,
    },
    Custom {
        message: String,
        span: Span,
    },
    AmbiguousModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
    MissingModule {
        module_name: &'static str,
        module_name_span: Span,
        filenames: Vec<String>,
    },
}

impl Diag for Error {
    fn message(&self) -> Cow<str> {
        match self {
            Self::ExpectedFound {
                expected,
                found,
                span: _,
            } => if expected.is_empty() {
                "Unexpected end of file".to_string()
            } else if expected.len() == 1 {
                format!(
                    "Expected {}, but found {}",
                    expected[0].green(),
                    found.as_deref().unwrap_or("something else")
                )
            } else {
                format!(
                    "Expected one of {}, but found {}",
                    expected
                        .iter()
                        .map(|expected| expected.green().to_string())
                        .collect::<Vec<_>>()
                        .join(", "),
                    found.as_deref().unwrap_or("something else")
                )
            }
            .into(),
            Self::Custom { message, span: _ } => message.into(),
            Self::AmbiguousModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Ambiguous module name".into(),
            Self::MissingModule {
                module_name: _,
                module_name_span: _,
                filenames: _,
            } => "Missing module".into(),
        }
    }

    fn spans(&self) -> Vec<ErrorSpan> {
        match self {
            Self::ExpectedFound {
                expected: _,
                found,
                span,
            } => found.as_ref().map_or_else(
                || vec![ErrorSpan::primary(*span)],
                |found| vec![ErrorSpan::primary_message(format!("Found {found}"), *span)],
            ),
            Self::Custom { message: _, span } => vec![ErrorSpan::primary(*span)],
            Self::AmbiguousModule {
                module_name: _,
                module_name_span,
                filenames: _,
            }
            | Self::MissingModule {
                module_name: _,
                module_name_span,
                filenames: _,
            } => vec![ErrorSpan::primary_message(
                "Declared here",
                *module_name_span,
            )],
        }
    }

    fn notes(&self) -> Vec<String> {
        match self {
            Self::AmbiguousModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "There exist multiple possible files for the module `{module_name}`:"
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::MissingModule {
                module_name,
                module_name_span: _,
                filenames,
            } => core::iter::once(format!(
                "Could not find a file for the module `{module_name}`"
            ))
            .chain(core::iter::once(
                "These are the possible files for this module:".into(),
            ))
            .chain(filenames.iter().map(|filename| format!("  - {filename}")))
            .collect(),
            Self::ExpectedFound { .. } | Self::Custom { .. } => {
                vec![]
            }
        }
    }

    fn kind(&self) -> Severity {
        Severity::Error
    }
}

pub fn convert(error: &Rich<impl Display, Span, &str>) -> Vec<Error> {
    fn convert_inner(reason: &RichReason<impl Display, &str>, span: Span) -> Vec<Error> {
        match reason {
            RichReason::ExpectedFound { expected, found } => vec![Error::ExpectedFound {
                expected: expected.iter().map(ToString::to_string).collect(),
                found: found.as_ref().map(|f| f.to_string()),
                span,
            }],
            RichReason::Custom(message) => vec![Error::Custom {
                message: message.to_owned(),
                span,
            }],
            RichReason::Many(reasons) => reasons
                .iter()
                .flat_map(|r| convert_inner(r, span))
                .collect(),
        }
    }

    convert_inner(error.reason(), *error.span())
}
