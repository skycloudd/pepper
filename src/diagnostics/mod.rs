use crate::span::Span;
use codespan_reporting::diagnostic::{LabelStyle, Severity};
use std::borrow::Cow;

pub mod error;
pub mod report;

pub trait Diag {
    fn message(&self) -> Cow<str>;
    fn spans(&self) -> Vec<ErrorSpan>;
    fn notes(&self) -> Vec<String>;
    fn kind(&self) -> Severity;
}

#[derive(Debug)]
pub struct ErrorSpan {
    message: Option<String>,
    span: Span,
    label_style: LabelStyle,
}

impl ErrorSpan {
    pub const fn primary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub fn primary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Primary,
        }
    }

    pub const fn secondary(span: Span) -> Self {
        Self {
            message: None,
            span,
            label_style: LabelStyle::Secondary,
        }
    }

    pub fn secondary_message(message: impl Into<String>, span: Span) -> Self {
        Self {
            message: Some(message.into()),
            span,
            label_style: LabelStyle::Secondary,
        }
    }
}
