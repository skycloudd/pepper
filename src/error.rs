use crate::ir::FileId;
use chumsky::{
    error::{Rich, RichReason},
    span::SimpleSpan,
};
use core::ops::Range;

#[derive(Clone, Debug)]
pub enum Error {
    ExpectedFound {
        expected: Vec<String>,
        found: Option<String>,
        span: Range<usize>,
    },
    Custom {
        message: String,
        span: Range<usize>,
    },
    Many(Vec<Error>),
}

fn convert_inner(reason: &RichReason<String, &str>, span: Range<usize>) -> Error {
    match reason {
        RichReason::ExpectedFound { expected, found } => Error::ExpectedFound {
            expected: expected.iter().map(ToString::to_string).collect(),
            found: found.as_ref().map(|f| f.to_string()),
            span,
        },
        RichReason::Custom(message) => Error::Custom {
            message: message.clone(),
            span,
        },
        RichReason::Many(reasons) => Error::Many(
            reasons
                .iter()
                .map(|r| convert_inner(r, span.clone()))
                .collect(),
        ),
    }
}

impl From<Rich<'_, String, SimpleSpan<usize, FileId<'_>>, &str>> for Error {
    fn from(error: Rich<String, SimpleSpan<usize, FileId<'_>>, &str>) -> Self {
        convert_inner(error.reason(), error.span().start..error.span().end)
    }
}

impl From<&Rich<'_, String, SimpleSpan<usize, FileId<'_>>, &str>> for Error {
    fn from(error: &Rich<String, SimpleSpan<usize, FileId<'_>>, &str>) -> Self {
        convert_inner(error.reason(), error.span().start..error.span().end)
    }
}
