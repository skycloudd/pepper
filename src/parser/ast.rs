use crate::{lexer::tokens::Interned, span::Spanned};

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Ast(pub Vec<Spanned<Item>>);

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Item {
    Function(Spanned<Function>),
    Import(Spanned<Path>),
    Struct(Spanned<Struct>),
    Enum(Spanned<Enum>),
    Module(Spanned<Module>),
}

#[allow(clippy::module_name_repetitions)]
pub type AstType = Type<Path>;

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Path {
    pub base: Spanned<Interned>,
    pub segments: Vec<Spanned<Interned>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Module {
    File(Spanned<Interned>),
    Submodule {
        name: Spanned<Interned>,
        ast: Spanned<Ast>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Function {
    pub name: Spanned<Interned>,
    pub params: Spanned<Vec<Spanned<FunctionParam>>>,
    pub return_ty: Option<Spanned<AstType>>,
    pub body: Spanned<Block>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct FunctionParam {
    pub name: Spanned<Interned>,
    pub ty: Spanned<AstType>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Struct {
    pub name: Spanned<Interned>,
    pub fields: Spanned<Vec<Spanned<StructField>>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct StructField {
    pub name: Spanned<Interned>,
    pub ty: Spanned<AstType>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Enum {
    pub name: Spanned<Interned>,
    pub variants: Spanned<Vec<Spanned<EnumVariant>>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum EnumVariant {
    Unit(Spanned<Interned>),
    Tuple {
        name: Spanned<Interned>,
        fields: Spanned<Vec<Spanned<AstType>>>,
    },
    Struct {
        name: Spanned<Interned>,
        fields: Spanned<Vec<Spanned<StructField>>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Statement {
    Expression(Spanned<Expression>),
    VarDecl {
        name: Spanned<Interned>,
        ty: Option<Spanned<AstType>>,
        value: Spanned<Expression>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Expression {
    Int(Spanned<Interned>),
    Float(Spanned<Interned>),
    Bool(Spanned<Interned>),
    String(Spanned<Interned>),
    Name(Spanned<Path>),
    Block(Spanned<Box<Block>>),
    Tuple(Spanned<Vec<Spanned<Self>>>),
    List(Spanned<Vec<Spanned<Self>>>),
    For {
        pattern: Spanned<Box<Pattern>>,
        iter: Spanned<Box<Self>>,
        body: Spanned<Box<Block>>,
    },
    BinaryOp {
        op: Spanned<BinaryOp>,
        lhs: Spanned<Box<Self>>,
        rhs: Spanned<Box<Self>>,
    },
    UnaryOp {
        op: Spanned<UnaryOp>,
        expr: Spanned<Box<Self>>,
    },
    Call {
        callee: Spanned<Box<Self>>,
        args: Spanned<Vec<Spanned<Self>>>,
    },
    Match {
        expr: Spanned<Box<Self>>,
        arms: Spanned<Vec<Spanned<MatchArm>>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Block {
    pub statements: Vec<Spanned<Statement>>,
    pub return_expr: Option<Spanned<Expression>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct MatchArm {
    pub pattern: Spanned<Pattern>,
    pub body: Spanned<Expression>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Pattern {
    pub pattern_type: Spanned<PatternType>,
    pub condition: Option<Spanned<Expression>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum PatternType {
    Name(Spanned<Path>),
    Int(Spanned<Interned>),
    Float(Spanned<Interned>),
    Bool(Spanned<Interned>),
    String(Spanned<Interned>),
    Tuple(Spanned<Vec<Spanned<Pattern>>>),
    List(Spanned<Vec<Spanned<ListPattern>>>),
    TupleType {
        name: Spanned<Path>,
        fields: Spanned<Vec<Spanned<Pattern>>>,
    },
    StructType {
        name: Spanned<Path>,
        fields: Spanned<Vec<Spanned<StructPatternField>>>,
    },
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct StructPatternField {
    pub name: Spanned<Interned>,
    pub pattern: Option<Spanned<Pattern>>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ListPattern {
    Pattern(Spanned<Pattern>),
    Rest,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    LessEquals,
    GreaterEquals,
    Less,
    Greater,
    Equals,
    NotEquals,
}

impl core::fmt::Display for BinaryOp {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Add => write!(f, "+"),
            Self::Sub => write!(f, "-"),
            Self::Mul => write!(f, "*"),
            Self::Div => write!(f, "/"),
            Self::LessEquals => write!(f, "<="),
            Self::GreaterEquals => write!(f, ">="),
            Self::Less => write!(f, "<"),
            Self::Greater => write!(f, ">"),
            Self::Equals => write!(f, "=="),
            Self::NotEquals => write!(f, "!="),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum UnaryOp {
    Neg,
    Not,
}

impl core::fmt::Display for UnaryOp {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Neg => write!(f, "-"),
            Self::Not => write!(f, "!"),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Type<P> {
    Primitive(Spanned<P>),
    Tuple(Spanned<Vec<Spanned<Self>>>),
    Never,
    Function {
        params: Spanned<Vec<Spanned<Self>>>,
        return_ty: Spanned<Box<Self>>,
    },
}

impl<P: core::fmt::Display> core::fmt::Display for Type<P> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Primitive(primitive) => write!(f, "{}", primitive.0),
            Self::Tuple(inner) => {
                write!(f, "(")?;
                for (i, ty) in inner.iter().enumerate() {
                    if i != 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{}", ty.0)?;
                }
                write!(f, ")")
            }
            Self::Never => write!(f, "!"),
            Self::Function { params, return_ty } => {
                write!(f, "func (")?;
                for (i, param) in params.iter().enumerate() {
                    if i != 0 {
                        write!(f, ", ")?;
                    }
                    write!(f, "{}", param.0)?;
                }
                write!(f, ") -> {}", return_ty.0)
            }
        }
    }
}
