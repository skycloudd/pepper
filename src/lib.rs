use camino::{Utf8Path, Utf8PathBuf};
use chumsky::{input::Input as _, span::Span as _, Parser as _};
use codespan_reporting::files::SimpleFiles;
use diagnostics::error::{convert, Error};
use lasso::ThreadedRodeo;
use parser::ast::{Ast, Item};
use span::{FileId, Span};
use std::sync::LazyLock;

pub mod diagnostics;
pub mod lexer;
pub mod parser;
pub mod scopes;
pub mod span;

static RODEO: LazyLock<ThreadedRodeo> = LazyLock::new(ThreadedRodeo::new);

#[must_use]
pub fn parse_file(
    filename: &Utf8Path,
    files: &mut SimpleFiles<Utf8PathBuf, String>,
    errors: &mut Vec<Error>,
) -> Option<Ast> {
    let source = std::fs::read_to_string(filename).unwrap();
    let file_id = FileId::new(files.add(filename.to_owned(), source.clone())).unwrap();

    let (tokens, lexer_errors) = lexer::lexer()
        .parse(source.with_context(file_id))
        .into_output_errors();

    errors.extend(lexer_errors.iter().flat_map(|error| convert(error)));

    let (mut ast, parser_errors) = tokens.as_ref().map_or_else(
        || (None, vec![]),
        |tokens| {
            let eoi = tokens
                .last()
                .map_or_else(|| Span::zero(file_id), |(_, span)| span.to_end());

            parser::parser()
                .parse(tokens.spanned(eoi))
                .into_output_errors()
        },
    );

    errors.extend(parser_errors.iter().flat_map(|error| convert(error)));

    if let Some(ast) = ast.as_mut() {
        insert_explicit_submodules(ast, filename, files, errors);
    }

    ast
}

pub fn insert_explicit_submodules(
    ast: &mut Ast,
    ast_filename: &Utf8Path,
    files: &mut SimpleFiles<Utf8PathBuf, String>,
    errors: &mut Vec<Error>,
) {
    ast.0
        .iter_mut()
        .filter_map(|item| match &mut item.0 {
            Item::Module(submodule) => Some(submodule),
            _ => None,
        })
        .filter(|submodule| submodule.ast.is_none())
        .for_each(|submodule| {
            let filenames = {
                let directory = ast_filename
                    .parent()
                    .unwrap()
                    .join(submodule.name.resolve());

                let submodule_filename = format!("{}.pr", submodule.name.resolve());

                [
                    ast_filename.with_file_name(&submodule_filename),
                    directory.join("mod.pr"),
                    directory.join(submodule_filename),
                ]
            };

            let mut existing = filenames.iter().filter(|filename| filename.exists());

            let existing_cloned = existing.clone();

            let filename = existing.next().unwrap_or_else(|| todo!());

            if existing.next().is_some() {
                errors.push(Error::AmbiguousModule {
                    module_name: submodule.name.resolve(),
                    module_name_span: submodule.name.1,
                    filenames: existing_cloned.map(ToString::to_string).collect(),
                });

                return;
            }

            let ast = parse_file(filename, files, errors);

            submodule.ast = ast;
        });
}

#[cfg(test)]
mod tests {
    use crate::parse_file;
    use camino::Utf8PathBuf;
    use codespan_reporting::files::SimpleFiles;
    use owo_colors::OwoColorize as _;
    use std::{io::Write as _, path::PathBuf};

    #[test]
    fn parser_tests() {
        insta::glob!("../tests/parser", "**/*.pr", |path| {
            let path = Utf8PathBuf::from_path_buf(path.to_path_buf()).unwrap();

            let short_path = path
                .components()
                .skip_while(|c| c.as_os_str() != "tests")
                .collect::<PathBuf>();

            insta::elog!("{} {} ...", "parsing".cyan(), short_path.display());

            let mut files = SimpleFiles::new();
            let mut errors = vec![];
            let ast = parse_file(&path, &mut files, &mut errors);

            assert!(errors.is_empty(), "errors: {errors:#?}");

            let snapshot_suffix = short_path
                .components()
                .map(|c| c.as_os_str().to_string_lossy())
                .collect::<Vec<_>>()
                .join("_");

            insta::with_settings!({
                description => std::fs::read_to_string(path).unwrap().trim(),
                snapshot_suffix => snapshot_suffix,
            }, {
                insta::assert_yaml_snapshot!(ast.unwrap());
            });
        });
    }
}
