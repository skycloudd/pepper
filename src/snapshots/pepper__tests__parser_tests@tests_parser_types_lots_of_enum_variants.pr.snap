---
source: src/lib.rs
description: "enum LotsOfEnumVariants =\n    | A\n    | B\n    | C\n    | D\n    | E\n    | F\n    | G\n    | H\n    | I\n    | J\n    | K\n    | L\n    | M\n    | N\n    | O\n    | P\n    | Q\n    | R\n    | S\n    | T\n    | U\n    | V\n    | W\n    | X\n    | Y\n    | Z\n    | AA\n    | BB\n    | CC\n    | DD\n    | EE\n    | FF\n    | GG\n    | HH\n    | II\n    | JJ\n    | KK\n    | LL\n    | MM\n    | NN\n    | OO\n    | PP\n    | QQ\n    | RR\n    | SS\n    | TT\n    | UU\n    | VV\n    | WW\n    | XX\n    | YY\n    | ZZ\n    | AAA\n    | BBB\n    | CCC\n    | DDD\n    | EEE\n    | FFF\n    | GGG\n    | HHH\n    | III\n    | JJJ\n    | KKK\n    | LLL\n    | MMM\n    | NNN\n    | OOO\n    | PPP\n    | QQQ\n    | RRR\n    | SSS\n    | TTT\n    | UUU\n    | VVV\n    | WWW\n    | XXX\n    | YYY\n    | ZZZ\n    | AAAA\n    | BBBB\n    | CCCC\n    | DDDD\n    | EEEE\n    | FFFF\n    | GGGG\n    | HHHH\n    | IIII\n    | JJJJ\n    | KKKK\n    | LLLL\n    | MMMM\n    | NNNN\n    | OOOO\n    | PPPP\n    | QQQQ\n    | RRRR\n    | SSSS\n    | TTTT\n    | UUUU\n    | VVVV\n    | WWWW\n    | XXXX\n    | YYYY\n    | ZZZZ\n    | AAAAA\n    | BBBBB\n    | CCCCC\n    | DDDDD\n    | EEEEE\n    | FFFFF\n    | GGGGG\n    | HHHHH\n    | IIIII\n    | JJJJJ\n    | KKKKK\n    | LLLLL\n    | MMMMM\n    | NNNNN\n    | OOOOO\n    | PPPPP\n    | QQQQQ\n    | RRRRR\n    | SSSSS\n    | TTTTT\n    | UUUUU\n    | VVVVV\n    | WWWWW\n    | XXXXX\n    | YYYYY\n    | ZZZZZ\n    | AAAAAA\n    | BBBBBB\n    | CCCCCC\n    | DDDDDD\n    | EEEEEE\n    | FFFFFF\n    | GGGGGG\n    | HHHHHH\n    | IIIIII\n    | JJJJJJ\n    | KKKKKK\n    | LLLLLL\n    | MMMMMM\n    | NNNNNN\n    | OOOOOO\n    | PPPPPP\n    | QQQQQQ\n    | RRRRRR\n    | SSSSSS\n    | TTTTTT\n    | UUUUUU\n    | VVVVVV\n    | WWWWWW\n    | XXXXXX\n    | YYYYYY\n    | ZZZZZZ\n    | AAAAAAA\n    | BBBBBBB\n    | CCCCCCC\n    | DDDDDDD\n    | EEEEEEE\n    | FFFFFFF\n    | GGGGGGG\n    | HHHHHHH\n    | IIIIIII\n    | JJJJJJJ\n    | KKKKKKK\n    | LLLLLLL\n    | MMMMMMM\n    | NNNNNNN\n    | OOOOOOO\n    | PPPPPPP\n    | QQQQQQQ\n    | RRRRRRR\n    | SSSSSSS\n    | TTTTTTT\n    | UUUUUUU\n    | VVVVVVV\n    | WWWWWWW\n    | XXXXXXX\n    | YYYYYYY\n    | ZZZZZZZ\n    | AAAAAAAA\n    | BBBBBBBB\n    | CCCCCCCC\n    | DDDDDDDD\n    | EEEEEEEE\n    | FFFFFFFF\n    | GGGGGGGG\n    | HHHHHHHH\n    | IIIIIIII\n    | JJJJJJJJ\n    | KKKKKKKK\n    | LLLLLLLL\n    | MMMMMMMM\n    | NNNNNNNN\n    | OOOOOOOO\n    | PPPPPPPP\n    | QQQQQQQQ\n    | RRRRRRRR\n    | SSSSSSSS\n    | TTTTTTTT\n    | UUUUUUUU\n    | VVVVVVVV\n    | WWWWWWWW\n    | XXXXXXXX\n    | YYYYYYYY\n    | ZZZZZZZZ\n    | AAAAAAAAA\n    | BBBBBBBBB\n    | CCCCCCCCC\n    | DDDDDDDDD\n    | EEEEEEEEE\n    | FFFFFFFFF\n    | GGGGGGGGG\n    | HHHHHHHHH\n    | IIIIIIIII\n    | JJJJJJJJJ\n    | KKKKKKKKK\n    | LLLLLLLLL\n    | MMMMMMMMM\n    | NNNNNNNNN\n    | OOOOOOOOO\n    | PPPPPPPPP\n    | QQQQQQQQQ\n    | RRRRRRRRR\n    | SSSSSSSSS\n    | TTTTTTTTT\n    | UUUUUUUUU\n    | VVVVVVVVV\n    | WWWWWWWWW\n    | XXXXXXXXX\n    | YYYYYYYYY\n    | ZZZZZZZZZ\n    | AAAAAAAAAA\n    | BBBBBBBBBB\n    | CCCCCCCCCC\n    | DDDDDDDDDD\n    | EEEEEEEEEE\n    | FFFFFFFFFF\n    | GGGGGGGGGG\n    | HHHHHHHHHH\n    | IIIIIIIIII\n    | JJJJJJJJJJ\n    | KKKKKKKKKK\n    | LLLLLLLLLL\n    | MMMMMMMMMM\n    | NNNNNNNNNN\n    | OOOOOOOOOO\n    | PPPPPPPPPP\n    | QQQQQQQQQQ\n    | RRRRRRRRRR\n    | SSSSSSSSSS\n    | TTTTTTTTTT\n    | UUUUUUUUUU\n    | VVVVVVVVVV\n    | WWWWWWWWWW\n    | XXXXXXXXXX\n    | YYYYYYYYYY\n    | ZZZZZZZZZZ"
expression: ast.unwrap()
input_file: tests/parser/types/lots_of_enum_variants.pr
snapshot_kind: text
---
- Enum:
    name: LotsOfEnumVariants
    variants:
      - Unit: A
      - Unit: B
      - Unit: C
      - Unit: D
      - Unit: E
      - Unit: F
      - Unit: G
      - Unit: H
      - Unit: I
      - Unit: J
      - Unit: K
      - Unit: L
      - Unit: M
      - Unit: N
      - Unit: O
      - Unit: P
      - Unit: Q
      - Unit: R
      - Unit: S
      - Unit: T
      - Unit: U
      - Unit: V
      - Unit: W
      - Unit: X
      - Unit: Y
      - Unit: Z
      - Unit: AA
      - Unit: BB
      - Unit: CC
      - Unit: DD
      - Unit: EE
      - Unit: FF
      - Unit: GG
      - Unit: HH
      - Unit: II
      - Unit: JJ
      - Unit: KK
      - Unit: LL
      - Unit: MM
      - Unit: NN
      - Unit: OO
      - Unit: PP
      - Unit: QQ
      - Unit: RR
      - Unit: SS
      - Unit: TT
      - Unit: UU
      - Unit: VV
      - Unit: WW
      - Unit: XX
      - Unit: YY
      - Unit: ZZ
      - Unit: AAA
      - Unit: BBB
      - Unit: CCC
      - Unit: DDD
      - Unit: EEE
      - Unit: FFF
      - Unit: GGG
      - Unit: HHH
      - Unit: III
      - Unit: JJJ
      - Unit: KKK
      - Unit: LLL
      - Unit: MMM
      - Unit: NNN
      - Unit: OOO
      - Unit: PPP
      - Unit: QQQ
      - Unit: RRR
      - Unit: SSS
      - Unit: TTT
      - Unit: UUU
      - Unit: VVV
      - Unit: WWW
      - Unit: XXX
      - Unit: YYY
      - Unit: ZZZ
      - Unit: AAAA
      - Unit: BBBB
      - Unit: CCCC
      - Unit: DDDD
      - Unit: EEEE
      - Unit: FFFF
      - Unit: GGGG
      - Unit: HHHH
      - Unit: IIII
      - Unit: JJJJ
      - Unit: KKKK
      - Unit: LLLL
      - Unit: MMMM
      - Unit: NNNN
      - Unit: OOOO
      - Unit: PPPP
      - Unit: QQQQ
      - Unit: RRRR
      - Unit: SSSS
      - Unit: TTTT
      - Unit: UUUU
      - Unit: VVVV
      - Unit: WWWW
      - Unit: XXXX
      - Unit: YYYY
      - Unit: ZZZZ
      - Unit: AAAAA
      - Unit: BBBBB
      - Unit: CCCCC
      - Unit: DDDDD
      - Unit: EEEEE
      - Unit: FFFFF
      - Unit: GGGGG
      - Unit: HHHHH
      - Unit: IIIII
      - Unit: JJJJJ
      - Unit: KKKKK
      - Unit: LLLLL
      - Unit: MMMMM
      - Unit: NNNNN
      - Unit: OOOOO
      - Unit: PPPPP
      - Unit: QQQQQ
      - Unit: RRRRR
      - Unit: SSSSS
      - Unit: TTTTT
      - Unit: UUUUU
      - Unit: VVVVV
      - Unit: WWWWW
      - Unit: XXXXX
      - Unit: YYYYY
      - Unit: ZZZZZ
      - Unit: AAAAAA
      - Unit: BBBBBB
      - Unit: CCCCCC
      - Unit: DDDDDD
      - Unit: EEEEEE
      - Unit: FFFFFF
      - Unit: GGGGGG
      - Unit: HHHHHH
      - Unit: IIIIII
      - Unit: JJJJJJ
      - Unit: KKKKKK
      - Unit: LLLLLL
      - Unit: MMMMMM
      - Unit: NNNNNN
      - Unit: OOOOOO
      - Unit: PPPPPP
      - Unit: QQQQQQ
      - Unit: RRRRRR
      - Unit: SSSSSS
      - Unit: TTTTTT
      - Unit: UUUUUU
      - Unit: VVVVVV
      - Unit: WWWWWW
      - Unit: XXXXXX
      - Unit: YYYYYY
      - Unit: ZZZZZZ
      - Unit: AAAAAAA
      - Unit: BBBBBBB
      - Unit: CCCCCCC
      - Unit: DDDDDDD
      - Unit: EEEEEEE
      - Unit: FFFFFFF
      - Unit: GGGGGGG
      - Unit: HHHHHHH
      - Unit: IIIIIII
      - Unit: JJJJJJJ
      - Unit: KKKKKKK
      - Unit: LLLLLLL
      - Unit: MMMMMMM
      - Unit: NNNNNNN
      - Unit: OOOOOOO
      - Unit: PPPPPPP
      - Unit: QQQQQQQ
      - Unit: RRRRRRR
      - Unit: SSSSSSS
      - Unit: TTTTTTT
      - Unit: UUUUUUU
      - Unit: VVVVVVV
      - Unit: WWWWWWW
      - Unit: XXXXXXX
      - Unit: YYYYYYY
      - Unit: ZZZZZZZ
      - Unit: AAAAAAAA
      - Unit: BBBBBBBB
      - Unit: CCCCCCCC
      - Unit: DDDDDDDD
      - Unit: EEEEEEEE
      - Unit: FFFFFFFF
      - Unit: GGGGGGGG
      - Unit: HHHHHHHH
      - Unit: IIIIIIII
      - Unit: JJJJJJJJ
      - Unit: KKKKKKKK
      - Unit: LLLLLLLL
      - Unit: MMMMMMMM
      - Unit: NNNNNNNN
      - Unit: OOOOOOOO
      - Unit: PPPPPPPP
      - Unit: QQQQQQQQ
      - Unit: RRRRRRRR
      - Unit: SSSSSSSS
      - Unit: TTTTTTTT
      - Unit: UUUUUUUU
      - Unit: VVVVVVVV
      - Unit: WWWWWWWW
      - Unit: XXXXXXXX
      - Unit: YYYYYYYY
      - Unit: ZZZZZZZZ
      - Unit: AAAAAAAAA
      - Unit: BBBBBBBBB
      - Unit: CCCCCCCCC
      - Unit: DDDDDDDDD
      - Unit: EEEEEEEEE
      - Unit: FFFFFFFFF
      - Unit: GGGGGGGGG
      - Unit: HHHHHHHHH
      - Unit: IIIIIIIII
      - Unit: JJJJJJJJJ
      - Unit: KKKKKKKKK
      - Unit: LLLLLLLLL
      - Unit: MMMMMMMMM
      - Unit: NNNNNNNNN
      - Unit: OOOOOOOOO
      - Unit: PPPPPPPPP
      - Unit: QQQQQQQQQ
      - Unit: RRRRRRRRR
      - Unit: SSSSSSSSS
      - Unit: TTTTTTTTT
      - Unit: UUUUUUUUU
      - Unit: VVVVVVVVV
      - Unit: WWWWWWWWW
      - Unit: XXXXXXXXX
      - Unit: YYYYYYYYY
      - Unit: ZZZZZZZZZ
      - Unit: AAAAAAAAAA
      - Unit: BBBBBBBBBB
      - Unit: CCCCCCCCCC
      - Unit: DDDDDDDDDD
      - Unit: EEEEEEEEEE
      - Unit: FFFFFFFFFF
      - Unit: GGGGGGGGGG
      - Unit: HHHHHHHHHH
      - Unit: IIIIIIIIII
      - Unit: JJJJJJJJJJ
      - Unit: KKKKKKKKKK
      - Unit: LLLLLLLLLL
      - Unit: MMMMMMMMMM
      - Unit: NNNNNNNNNN
      - Unit: OOOOOOOOOO
      - Unit: PPPPPPPPPP
      - Unit: QQQQQQQQQQ
      - Unit: RRRRRRRRRR
      - Unit: SSSSSSSSSS
      - Unit: TTTTTTTTTT
      - Unit: UUUUUUUUUU
      - Unit: VVVVVVVVVV
      - Unit: WWWWWWWWWW
      - Unit: XXXXXXXXXX
      - Unit: YYYYYYYYYY
      - Unit: ZZZZZZZZZZ
